restore-sql-server-transaction-logs/Function/main.py

```diff
@@ -0,0 +1,274 @@
+                    object_name,
+                    if_generation_match=None
+                )
+
+            if blob_copy is None:
+                logger.info("Could not copy to processed bucket.")
+
+        except Exception as err:
+            logger.info(f"Could not copy the object {object_name} to the processed bucket. Error: {err.args}")
+
+# Triggered by a change in a storage bucket
+@functions_framework.cloud_event
+def fn_restore_log(cloud_event):
+    data = cloud_event.data
+    
+    source_bucket_name = data["bucket"]
+    object_name = data["name"]
+    
+    logging.basicConfig(level=logging.INFO)
+    logger = logging.getLogger()
+    
+    if object_name.endswith('/'):
+        logger.info("Folder creation encountererd. Operation skipped.")
+        return "Operation skipped."
+
+    valid_env_v, msg, processed_bucket_name, use_fixed_file_name_format, max_request_attempts, max_request_fetch_time_seconds, max_operation_fetch_time_seconds = get_environment_variables()
+
+    if not(valid_env_v):        
+        raise RuntimeError(f"Operation failed. Environment variables validation failed: {msg}.")        
+    
+    if (use_fixed_file_name_format.lower()=="true"):        
+        logger.info("Using fixed file name format")
+        valid_file_name, msg, csql_instance_name, database_name, backup_type, no_recovery = extract_info_from_file_name(object_name)
+
+        if not(valid_file_name):
+            raise RuntimeError(f"Operation failed. File Name validation failed: {msg}.")
+    else:
+        logger.info(f"Using metadata")
+        valid_metadata, msg, csql_instance_name, database_name, backup_type, no_recovery = extract_info_from_file_metadata(source_bucket_name, object_name)        
+
+        if not(valid_metadata):
+            raise RuntimeError(f"Operation failed. Metadta validation failed: {msg}.")
+
+    request_attempts = 0
+    log_file_processed = False
+
+    creds, project_id = google.auth.default()
+    auth_req = google.auth.transport.requests.Request()
+    creds.refresh(auth_req)
+
+    service = googleapiclient.discovery.build('sqladmin', 'v1beta4', credentials=creds)
+
+    storage_client = storage.Client()
+    source_bucket = storage_client.bucket(source_bucket_name)    
+
+    import_context_body = get_import_context(source_bucket_name,object_name,database_name,backup_type, no_recovery)
+
+    while log_file_processed == False:
+
+        request_attempts += 1
+
+        try:
+            
+            logger.info(f"Attempt to process object {object_name}. Try {request_attempts} out of {max_request_attempts}...")
+
+            request = service.instances().import_(project=project_id, instance=csql_instance_name, body=import_context_body)
+            response = request.execute()
+            
+            logger.info(f"Executed request: {import_context_body}")
+                        
+            operation_status_done = False
+            operation_status_fetch_attempts=0

```

Comment:
Where are you using this?

Reply:
It has to do with the inner loop (see comment above). We ask a few times for the status of the operation. 
When it is no longer in "pending" status, we defined it as done (can be succeeded or failed).

Reply:
Left only one loop that pools the status of the import operation periodically until its status is Done.

Reply:
Got it

Reply:
Can we close this?

Reply:
Yes.